nn/transformer.py

```diff
@@ -0,0 +1,231 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union
+from . import relu, layer_norm, dropout, LayerRef, Module, ModuleList, Linear
+
+
+class TransformerEncoderLayer(Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, drop: float = 0.1, activation=relu,
+               layer_norm_eps: float = 1e-5, norm_first: bool = False) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param drop: Dropout value, PyTorch name: dropout
+    :param activation: activation functional
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
```

Comment:
I don't understand the comment.

Reply:
Yeah the norm_first comment now reading it is somewhat weird, will update it

Reply:
I'm referring to the comment in this line here.

Reply:
these comments are internal, that this line will most likely change once AttentionModules are implemented, depending on what they require as input etc.

Reply:
By "internal" you mean still work-in-progress? Then write that in the comment. (TODO or WIP or so)
Although I would just remove the comment. The code could maybe just work like this.

Reply:
MultiheadAttention is just not implemented yet, so it is not clear what the input for that would be right now. I replaced the comment by: "to be implemented".